slot2ssa: remove `TypedSlot`

topolarity opened the pull request:
Dependent on https://github.com/JuliaLang/julia/pull/50924 .

This removes the last remaining usage of `TypedSlot`.

As a bonus, this evicts all of the "tmerge" type-iteration that was being done in slot2ssa.

TODO:
 - [x] use `bb_vartables` for `SLOT_USEDUNDEF`, so that we can remove `find_undefs!`
 - [x] confirm perf improvement with benchmarks

topolarity committed:
slot2ssa: Insert PiNodes based on `bb_vartables`

This is the last usage of TypedSlot (we should be clear now to remove
that step from the end of type inference now).

Compared to the prior scheme, this now inserts PiNodes at every PhiNode
corresponding to a given slot, rather than at each load. In principle,
that should mean that the types we compute across PhiNodes are more
precise than before.

I believe that means that all slot uses should be at least as precise
as before, but I haven't completely convinced myself yet.

topolarity committed:
compiler: Remove `TypedSlot`

Since this is now unused by `slot2ssa` downstream, we can happily remove
it from the rest of the compiler.

There's one more piece of logic that I still need to move downstream: we
should use the `bb_vartables` for undefined-ness as well. That will
allow us to finally remove the (now incorrectly named) `annotate_slot_load!`

topolarity committed:
slot2ssa: Remove type iteration

Now that we have the `bb_vartables` from Inference, we can get a
converged (and frequently more precise) result for our inserted PhiNodes
directly, instead of trying to compute the fix point in the optimizer.

This change also improves the way that we are inserting PiNodes to
ensure that they are present in every block that requires them for
precision, not just where we are already inserting PhiNodes. This is
necessary due to the presence of `Conditional`, which means we may
have a refined type in a block even when dominated by a definition
whose type was precise outside of the Conditional.

topolarity commented on base/compiler/ssair/slot2ssa.jl:
```diff
@@ -1,13 +1,5 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
-struct TypedSlot
```
Worth mentioning that deleting this definition breaks JuliaInterpreter (and therefore Revise)

Keno replied:
We'll put in a version check.

topolarity committed:
slot2ssa: Remove usage of `SLOT_USEDUNDEF`

Updates slot2ssa to use the `bb_vartables` for `undef` information.
As a bonus, evicts the final usage of `tmerge` in slot2ssa.

This also fixes a few crashes from the prior commit, esp. where I was
emitting PiNode(UndefToken(), Union{}) or where the PhiC nodes were
not being ordered correctly.

topolarity commented on base/compiler/typeinfer.jl:
```diff
@@ -704,52 +664,54 @@ function type_annotate!(interp::AbstractInterpreter, sv::InferenceState, run_opt
-    # finish marking used-undef variables
-    for j = 1:nslots
-        if undefs[j]
-            slotflags[j] |= SLOT_USEDUNDEF | SLOT_STATICUNDEF
```
The optimizer no longer relies on these flags, but will external consumers need them?

Keno replied:
We'll find out. I'm somewhat sceptical, but if they are, let's wean them off it.

topolarity commented:
I still need to fix up the test suite, but otherwise this is ready for review now.

Keno commented on test/compiler/inference.jl:
```diff
@@ -1961,31 +1961,32 @@ g26339(T) = T === Int ? 1 : ""
-    end
-    @test foundslot
-end
+# TODO: Re-write test
```
IR here is:
```
CodeInfo(
1 ─      nothing::Nothing
2 ┄ %2 = φ (#1 => (1, 1), #3 => %8)::Union{Nothing, Tuple{Int64, Int64}}
│   %3 = (%2 === Main.nothing)::Bool
│   %4 = Core.Intrinsics.not_int(%3)::Bool
└──      goto #4 if not %4
3 ─ %6 = π (%2, Tuple{Int64, Int64})
│   %7 = Core.getfield(%6, 2)::Int64
│   %8 = invoke Main.f25261(%7::Int64)::Union{Nothing, Tuple{Int64, Int64}}
└──      goto #2
4 ─      return nothing
) => Nothing

```
Let's search for the PiNode.

topolarity committed:
Inference: Always update IR to avoid unreachable jumps

This allows downstream consumers, like the optimizer, to be sure that
they are only visiting blocks that Inference also visited.

Also update tests and remove some unnused variables in `slot2ssa`.

topolarity committed:
test: Remove check for `GotoIfNot`

Now that type inference modifies branches to reflect static control flow
information, we can't search for the branch that we want to optimize away.

Pangoraw commented on base/compiler/ssair/slot2ssa.jl:
```diff
@@ -102,14 +94,12 @@ function fixup_slot!(ir::IRCode, ci::CodeInfo, idx::Int, slot::Int, stmt::UnoptS
     end
     if isa(stmt, SlotNumber)
```
I think this is now always true since `stmt::SlotNumber`, which means that `fixup_slot!` always returns `ssa` (even when `ssa === UNDEF_TOKEN`) and the `stmt` argument can be removed.

Pangoraw commented on base/compiler/ssair/slot2ssa.jl:
```diff
@@ -130,26 +120,17 @@ function fixemup!(@specialize(slot_filter), @specialize(rename_slot), ir::IRCode
         val = stmt.args[1]
-        if isa(val, UnoptSlot)
+        if isa(val, SlotNumber)
             slot = slot_id(val)
```
`slot` now seems unused.

```suggestion
```

Pangoraw reviewed (commented):
Minor comments from reading the PR.

For the slotflags validity, there is already `SLOT_ASSIGNEDONCE` and `SLOT_USED` ~~whose value seems to never be set (from my research in the codebase).~~

https://github.com/JuliaLang/julia/blob/fdd16bf0aa25aa1037d88c1cb4aa651cb2bdcce6/base/compiler/optimize.jl#L12

https://github.com/JuliaLang/julia/blob/fdd16bf0aa25aa1037d88c1cb4aa651cb2bdcce6/base/reflection.jl#L1244

Keno commented:
> For the slotflags validity, there is already `SLOT_ASSIGNEDONCE` and `SLOT_USED` whose value seems to never be set (from my research in the codebase).

Some of the slotflags are computed in the frontend

topolarity commented on base/compiler/typeinfer.jl:
```diff
@@ -682,7 +642,7 @@ function find_dominating_assignment(id::Int, idx::Int, sv::InferenceState)
 
 # annotate types of all symbols in AST, preparing for optimization
-function type_annotate!(interp::AbstractInterpreter, sv::InferenceState, run_optimizer::Bool)
+function type_annotate!(interp::AbstractInterpreter, sv::InferenceState)
```
I'm a bit uncertain about getting rid of this flag, but with these changes the meaning would have been "generate IR that is valid to pass into the optimizer" not just "perform a few optional optimizations".

We could also:
 - add more checks in slot2ssa to deal with `run_optimizer=false`, or else
 - annotate the dead branches for the optimizer to do this modification itself

Keno replied:
Meh, I think it's fine. It's possible we should just not run this function at all if the optimizer isn't being run, but I think this is fine as is.

aviatesk replied:
Can we move `5: replace this live GotoIfNot with:` to the optimizer pass rather than doing it unconditionally here? Then `type_annotate!` really fixes up inference-only types only, which sounds much cleaner.

topolarity replied:
Done!

topolarity replied:
A downside of this approach just occurred to me, which is that running type-inference and then using our [external entrypoint](https://github.com/JuliaLang/julia/pull/50943/files#diff-e04705952aec52ac6b24bd37c25cd11b8ba7d72b488b645bd5d2c104f79ccaabR143) for the optimizer is going be strictly worse than the result we obtain internally, since we won't be able to prune branches or insert PiNodes where we used to have `TypedSlot`.

I don't think there's a way around that, unless we can add this reachability/vartables information to `CodeInfo`?

aviatesk replied:
I think that's fine and I am even not sure if there are serious users of the external optimization entrypoint other than the Base test suite. If it turns out to be a problem, we can allow `OptimizationState(::MethodInstance, ::CodeInfo, ::AbstractInterpreter)` to accept optional reachability/variable information.

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -704,63 +664,52 @@ function type_annotate!(interp::AbstractInterpreter, sv::InferenceState, run_opt
-    # 2. mark used-undef slots (required by the `slot2reg` conversion)
-    # 3. mark unreached statements for a bulk code deletion (see issue #7836)
-    # 4. widen slot wrappers (`Conditional` and `MustAlias`) and remove `NOT_FOUND` from `ssavaluetypes`
+    # 1. mark used-undef slots (required by the `slot2reg` conversion)
+    # 2. mark unreached statements for a bulk code deletion (see issue #7836)
+    # 3. widen slot wrappers (`Conditional` and `MustAlias`) and remove `NOT_FOUND` from `ssavaluetypes`
```
```suggestion
    # 1. widen slot wrappers (`Conditional` and `MustAlias`) and remove `NOT_FOUND` from `ssavaluetypes`
```
since we no longer perform the bulk deletion here and need the used-undef marking.

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -704,63 +664,52 @@ function type_annotate!(interp::AbstractInterpreter, sv::InferenceState, run_opt
     #    NOTE because of this, `was_reached` will no longer be available after this point
-    # 5. eliminate GotoIfNot if either or both branches are statically unreachable
+    # 4. eliminate GotoIfNot if either or both branches are statically unreachable
     changemap = nothing # initialized if there is any dead region
```
```suggestion
```

aviatesk commented on base/compiler/ssair/slot2ssa.jl:
```diff
@@ -675,11 +633,18 @@ function construct_ssa!(ci::CodeInfo, ir::IRCode, domtree::DomTree,
                 insertpoint = first_insert_for_bb(code, cfg, li)
+                @assert sv.bb_vartables[li] !== nothing
+                vt = sv.bb_vartables[li][idx]
+                typ = widenslotwrapper(ignorelimited(vt.typ))
```
We may want to `widenslotwrapper` within `type_annotate!` and force `sv.bb_vartables` to be valid for `optimizer_lattice` always?

topolarity replied:
Yeah, great idea - Let's just widen the types at the inference boundary like we do elsewhere

aviatesk reviewed (commented):
Looks great. There are few potential improvements like cleaning up `type_annotate!` further by shifting the optimization of `GotoIfNot` to slot2reg, but the changes are very nice overall.

topolarity commented on base/show.jl:
```diff
@@ -1424,9 +1424,9 @@ show(io::IO, s::Symbol) = show_unquoted_quote_expr(io, s, 0, 0, 0)
 # While this isn’t true of ALL show methods, it is of all ASTs.
 
-using Core.Compiler: TypedSlot, UnoptSlot
+using Core.Compiler: SlotNumber
 
```
This seems to still be used though?

topolarity replied:
Oh I see, this is already exported in `Core` :+1:

topolarity committed:
inference: Move unreachable branch updates to `convert_to_ircode`

This means that inference modifies types and not IR statements
themselves. In order to make this happen, we pass have to pass
per-statement reachability information to the optimizer.

Also fixes up assorted review feedback.

topolarity committed:
inference: Wrap `unreachable` stmts in `Const(...)`

We were doing this before, but the previous commit had changed it. The
problem is that the optimizer cannot rely on the `unreachable` vector
for its external entrypoint.

This restores the existing logic so that type inference will give you a
result that you can pipe into the optimizer and have it work, although
suboptimally since we won't be able to prune any unreachable branches.

topolarity committed:
Move `unreachable` into InferenceState

I think it'd be a more maintainable data flow to have these results
collected manually and explicitly passed to the conversion routine for
the optimizer, but for now this follows the established pattern and
allows us to create a complete OptimizationState from an InferenceState.

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -523,12 +532,39 @@ function convert_to_ircode(ci::CodeInfo, sv::OptimizationState)
+    code = copy_exprargs(ci.code)
+    for i = 1:length(code)
+        expr = code[i]
+        if !(i in sv.unreachable) && isa(code[i], GotoIfNot)
```
```suggestion
        if !(i in sv.unreachable) && isa(expr, GotoIfNot)
```

aviatesk reviewed (commented):
We need to update `OptimizationState` calls within our tests since the `OptimizationState` signature has been updated. Alternatively we can make it optional argument `unreachable::BitSet=BitSet()`.

aviatesk committed:
Merge branch 'master' into rm-TypedSlot

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50943#issuecomment-1692528693) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c4a5f06_vs_4ac6b05/report.md).

aviatesk committed:
fix up tests

aviatesk committed:
perf improvements

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -130,12 +130,14 @@ mutable struct OptimizationState{Interp<:AbstractInterpreter}
     return OptimizationState(sv.linfo, sv.src, nothing, sv.stmt_info, sv.mod,
-                             sv.sptypes, sv.slottypes, inlining, sv.cfg, sv.insert_coverage)
+                             sv.sptypes, sv.slottypes, inlining, sv.cfg,
+                             sv.bb_vartables, sv.insert_coverage)
```
Ah, `bb_vartables` here may have `nothing` entry for unreachable blocks. Never mind.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/50943#issuecomment-1692755171) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/59b97ea_vs_4ac6b05/report.md).

aviatesk commented:
The benchmark looks great now. Let's merge.

aviatesk commented:
Great job @topolarity!

maleadt commented:
Lots of PkgEval breakage (500 packages crash now) that's probably related to this PR:

```
Internal error: encountered unexpected error in runtime:
MethodError(f=Base.string, args=(Expr(:call, :!==, :varstate, :nothing),), world=0x00000000000015b0)
jl_method_error_bare at /source/src/gf.c:2203
jl_method_error at /source/src/gf.c:2221
jl_lookup_generic_ at /source/src/gf.c:3052 [inlined]
ijl_apply_generic at /source/src/gf.c:3067
macro expansion at ./error.jl:231 [inlined]
construct_ssa! at ./compiler/ssair/slot2ssa.jl:752
slot2reg at ./compiler/optimize.jl:697
run_passes at ./compiler/optimize.jl:494
run_passes at ./compiler/optimize.jl:509 [inlined]
optimize at ./compiler/optimize.jl:458
jfptr_optimize_37748.1 at /opt/julia/lib/julia/sys.so (unknown line)
_jl_invoke at /source/src/gf.c:2870 [inlined]
ijl_apply_generic at /source/src/gf.c:3071
_typeinf at ./compiler/typeinfer.jl:265
typeinf at ./compiler/typeinfer.jl:212
```

vtjnash commented:
Could you look at why this seems to have broken performance of IdDict according to: https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/5e5416a_vs_4ac6b05/report.md

(in the future, it would be good to runbenchmarks with the specifier `ALL` before merging significant changes, though that is as much a reminder to myself as a suggestion to you)

topolarity commented:
> Could you look at why this seems to have broken performance of IdDict according to: https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/5e5416a_vs_4ac6b05/report.md

At least some of this appears to be due to increased precision on `PhiNode`s.

Previously in `setindex!(::IdDict{Int, Int}, ::Any, ::Any)` we were emitting:
```
%15 = φ (#4 => %14, #3 => val)::Any
...
%48 = π (%15, Int64)
```

Now we emit the improved:
```
%15 = φ (#4 => %14, #3 => val)::Int64
```

The problem is that codegen unboxes the refined PhiNode too eagerly and then has to re-box it for a function call
(kudos to @gbaraldi for identifying the problem)
